README.MD
# Dealabs Api Notifications

[![Tests](https://github.com/thomasync/dealabs-api-notifications/actions/workflows/python-package.yml/badge.svg)](https://github.com/thomasync/dealabs-api-notifications/actions/workflows/python-package.yml)

## Envoi des notifications sur téléphone pour les erreurs de prix Dealabs (via pushover)

1. Créer un compte sur [PushOver](https://pushover.net/)
2. Générer une [application](https://pushover.net/apps/build)
3. Renommer le fichier **.env.example** en **.env**
4. Remplir les valeurs _PUSHOVER_TOKEN_ et _PUSHOVER_USER_
5. Sur l'application : Settings -> Advanced Settings -> Auto-open URLs
6. Optionnel: Ajouter un fichier blacklist.txt où chaque ligne permet d'ignorer des deals (correspondance effectuée dans le titre)

Exemple de la commande:

```python
pip install -r requirements.txt

# python3 main.py [refresh_seconds] [minimum_discount] [free_products] [expire_notification] [open_dealabs] [priority_only_first] [free_products_priority]
python3 main.py 60 90 0 0 1 1 0
```

Optionnel dans le fichier .env:

- _REFRESH_SECONDS_ : Actualiser la recherche des deals toutes les x seconds (60s par défaut)
- _MIN_DISCOUNT_ : Représente le pourcentage minimum pour déclencher la notification (90% par défaut)
- _FREE_PRODUCTS_ : Capturer également les produits gratuits (Désactivé par défaut)
- _EXPIRE_NOTIFICATION_ : Prévient de l'expiration des tokens PushOver la veille pour les modifier (attention c'est basé sur la date de modification du fichier **.env**) (Désactivé par défaut)
- _OPEN_DEALABS_ : Au click sur la notification ouvrir Dealabs, au lieu du site du deal (Activé par défaut)
- _PRIORITY_ONLY_FIRST_ : Notifier uniquement en priorité a l'apparition du deal et pas à chaque mise à jour (Activé par défaut)
- _FREE_PRODUCTS_PRIORITY_ : Envoi également les notifications des produits gratuit en priorité (Désactivé par défaut)

---

🚨 Attention : **PushOver** n'est gratuit qu'un mois, il faut soit refaire un compte et changer les tokens une fois par mois, soit payer 5,99€ définitivement.

---

## Utilisation de la librairie de façon indépendante

1. Déplacer le dossier **dealabs** dans votre projet
2. Importer la classe Dealabs: `from dealabs import Dealabs`
3. Instanciez: `dealabs = Dealabs()`

### Exemples

#### Récuperer les deals avec au minimum 80% de réduction et sans les produits gratuits

```python
from dealabs import Dealabs
dealabs = Dealabs(80, 0)
deals = dealabs.getDeals()
```

#### Récuperer les deals avec au minimum 80% de réduction, avec les produits gratuits mais ne pas ajouter ceux en cours de modération

```python
from dealabs import Dealabs
dealabs = Dealabs(80, 1)
deals = dealabs.getDeals(False)
```

#### Récuper un deal depuis son id

```python
from dealabs import Dealabs
dealabs = Dealabs()
deal = dealabs.getDeal(2497616)
print(deal.json())
```

#### Récuper un deal depuis son url

```python
from dealabs import Dealabs
dealabs = Dealabs()
deal = dealabs.getDealFromUrl("2497616")
print(deal.json())
```

## [CHANGELOG](https://github.com/thomasync/dealabs-api-notifications/blob/main/CHANGELOG.MD)

2023 - [@thomasync](https://github.com/thomasync)


CHANGELOG.MD
# Changelog

Toutes les modifications notable de ce projet seront documentées dans ce fichier.

Le format est basé sur [Keep a Changelog](https://keepachangelog.com/fr/1.0.0/),
et ce projet adhère à [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [3.1.1] - 2023-02-19

### Corrigé

- La blacklist ne prenait pas en compte le nom du marchant

## [3.1.0] - 2023-01-28

### Ajouté

- Ajout d'un fichier blacklist.txt où chaque ligne correspond à des mots bannis du titre

### Corrigé

- Correction des tests getDeals() qui n'étaient pas correctement typé

## [3.0.1] - 2023-01-24

### Ajouté

- Ajout du fichiers requirements.txt pour les dépendances
- Ajout de Github Actions

## [3.0.0] - 2023-01-24

### Modifié

- Refactoring pour déplacer la logique métier dans la classe Dealabs et Deal
- Déplacement des outils et de gestion dans la classe Utils
- Repo renommé en dealabs-api-notifications

### Ajouté

- Ajout des tests pour vérifier la classe Dealabs et Deals ainsi que l'accès à l'api soit toujours fonctionnel
- Ajout du paramètre _FREE_PRODUCTS_PRIORITY_ pour pouvoir afficher les notifications des produits gratuits de façon non prioritaire et se concentrer sur les erreurs de prix

## [2.1.0] - 2023-01-21

### Ajouté

- Ajout du paramètre _PRIORITY_ONLY_FIRST_ pour notifier uniquement en priorité a l'apparition du deal et pas à chaque mise à jour
- Les deals envoyés sont en cache dans un fichier send.json

## [2.0.2] - 2023-01-21

### Corrigé

- Les deals designés comme locaux, mais de type nationaux n'étaient pas récupérés

## [2.0.1] - 2023-01-20

### Corrigé

- Catch les erreurs lors de la récupération des redirections d'urls dans getFinalUrl()

## [2.0.0] - 2023-01-19

### Modifié

- Refactoring complet pour tout inclure dans un module
- La récupération passe désormais par graphQL et plus l'api interne
- Les images ne sont plus récupérées via l'api des images, mais le CDN public

### Ajouté

- Ajout de la fonctionnalité _OPEN_DEALABS_ qui permet de choisir d'ouvrir l'application dealabs ou le site du deal
- Il est maintenant possible d'inclure le module pour l'utiliser directement en python
- Le point d'entrée (main.py) accepte maintenant des arguments [refresh_seconds] [minimum_discount] [free_products] [expire_notification] [open_dealabs]
- Retour du getDeal()

## [1.0.7] - 2023-01-18

### Ajouté

- Ajout du paramètre _FREE_PRODUCTS_ pour capturer les produits gratuits

## [1.0.6] - 2023-01-16

### Corrigé

- Changement de l'ordre des paramètres pour correspondre a l'oauth_signature

## [1.0.5] - 2023-01-16

### Modifié

- La requete pour récupérer les deals contient maintenant un compte fictif, ce qui permet de récupérer les deals plus rapidement

## [1.0.4] - 2023-01-15

### Modifié

- Refactorisation du code
- Ajout de timeouts à chaque requete

### Supprimé

- getProduct() a été supprimé car il ne fonctionnait pas, et était sécurisé avec un oauth_signature

## [1.0.3] - 2023-01-14

### Corrigé

- Ajout du prix, réduction, marchand et les groupes du deals au hash

## [1.0.2] - 2023-01-14

### Modifié

- Génération d'un hash au lieu de prendre l'id du deal, sinon les modifications ultérieures à la création du deal n'étaient pas prises en compte

## [1.0.1] - 2023-01-13

### Ajouté

- Ajout de la fonctionalité permettant d'être prevenu lorsque les tokens pushOver vont expirer

## [1.0.0] - 2023-01-13

Création du projet


tests/test_api.py
from dealabs import Dealabs, Utils
from unittest.mock import patch
import os


def test_getCookies():
    dealabs = Dealabs()
    cookies = dealabs.getCookies(False)

    assert cookies is not None
    assert "xsrf_t" in cookies
    assert "pepper_session" in cookies
    assert os.path.exists(Utils.getCacheFolder() + "cookies.json")

    cookies_from_cache = dealabs.getCookies(True)
    assert cookies_from_cache is not None
    assert cookies == cookies_from_cache


@patch('dealabs.Dealabs.getCookies')
def test_getDeals_without_cookies(mock_getCookies):
    mock_getCookies.return_value = []

    dealabs = Dealabs()
    deals = dealabs.getDeals(False)

    assert mock_getCookies.call_count == 3
    assert deals is not None
    assert len(deals) == 0


def test_getDeals_without_inModeration():
    dealabs = Dealabs()
    deals = dealabs.getDeals(False)

    assert deals is not None
    assert len(deals) > 0


def test_getDeals_with_inModeration():
    dealabs = Dealabs()
    deals = dealabs.getDeals(True)

    assert deals is not None
    assert len(deals) > 0


tests/test_dealabs.py
from dealabs import Dealabs, Deal
from unittest.mock import patch, MagicMock
import json


def check_deal_fields(deal):
    if type(deal) is Deal:
        deal = deal.json()

    assert "id" in deal
    assert "type" in deal
    assert "title" in deal
    assert "description" in deal
    assert "status" in deal
    assert "price" in deal
    assert "nextBestPrice" in deal
    assert "priceReduction" in deal
    assert "priceDiscount" in deal
    assert "image" in deal
    assert "redirectUrl" in deal
    assert "url" in deal
    assert "updatedAt" in deal
    assert "createdAt" in deal
    assert "publishedAt" in deal
    assert "merchant" in deal


def test_constructor_empty():
    dealabs = Dealabs()

    assert dealabs is not None
    assert dealabs.minimum_discount == 90
    assert dealabs.free_products == 0
    assert dealabs.expire_notification == 0
    assert dealabs.open_dealabs == 1
    assert dealabs.priority_only_first == 1
    assert dealabs.refresh_seconds == 0


def test_constructor_with_parameters():
    dealabs = Dealabs(10, 1, 1, 0, 0)

    assert dealabs is not None
    assert dealabs.minimum_discount == 10
    assert dealabs.free_products == 1
    assert dealabs.expire_notification == 1
    assert dealabs.open_dealabs == 0
    assert dealabs.priority_only_first == 0
    assert dealabs.refresh_seconds == 0


def test_setPushOver():
    dealabs = Dealabs()
    dealabs.setPushOver("token", "user")
    assert dealabs.token == "token"
    assert dealabs.user == "user"


@patch('dealabs.Dealabs._getThreads')
def test_getDeals_check_fields(mock_getThreads):
    with open('tests/mocks/deals.json') as file:
        mock_deals = json.load(file)

    mock_getThreads.return_value = mock_deals
    dealabs = Dealabs()

    deals = dealabs.getDeals(False)

    # Check ordered by updatedAt asc
    for i in range(0, len(deals) - 1):
        assert deals[i].updatedAt <= deals[i + 1].updatedAt

    # Check all fields
    for deal in deals:
        assert hasattr(deal, '_addedModeration')
        check_deal_fields(deal)


@patch('dealabs.Dealabs._getThreads')
def test_getDeals_check_deactivated(mock_getThreads):
    with open('tests/mocks/deals.json') as file:
        mock_deals = json.load(file)

    mock_getThreads.return_value = mock_deals
    dealabs = Dealabs()

    for deal in dealabs.getDeals(False):
        if deal.id == 2496398:
            assert deal.isDeactivated() == True
        elif deal.id == 2496417:
            assert deal.isDeactivated() == True


@patch('dealabs.Dealabs._getThreads')
def test_getDeals_check_deactivated_blacklist(mock_getThreads):
    with open('tests/mocks/deals.json') as file:
        mock_deals = json.load(file)

    mock_getThreads.return_value = mock_deals
    dealabs = Dealabs()
    dealabs.setBlacklist(["nintendo eshop", "ebook"])

    for deal in dealabs.getDeals(False):
        if deal.id == 496500:
            assert deal.isDeactivated() == True
        elif deal.id == 496501:
            assert deal.isDeactivated() == True


@patch('dealabs.Dealabs._getThreads')
def test_getDeals_check_location(mock_getThreads):
    with open('tests/mocks/deals.json') as file:
        mock_deals = json.load(file)

    mock_getThreads.return_value = mock_deals
    dealabs = Dealabs()

    for deal in dealabs.getDeals(False):
        # isLocal and not isNational
        if deal.id == 2496418:
            assert deal.isLocal() == True

        # not isLocal and not selectedLocations
        elif deal.id == 2496420:
            assert deal.isLocal() == False

        # isLocal and isNational
        elif deal.id == 2496421:
            assert deal.isLocal() == False


@patch('dealabs.Dealabs._getThreads')
def test_getDeals_check_error(mock_getThreads):
    with open('tests/mocks/deals.json') as file:
        mock_deals = json.load(file)

    mock_getThreads.return_value = mock_deals
    dealabs = Dealabs()

    for deal in dealabs.getDeals(False):
        # in Title
        if deal.id == 2496446:
            assert deal.isError() == True

        # in Description
        elif deal.id == 2496447:
            assert deal.isError() == True

        # in Groups
        elif deal.id == 2496443:
            assert deal.isError() == True


@patch('dealabs.Dealabs._getThreads')
def test_getDeals_check_free(mock_getThreads):
    with open('tests/mocks/deals.json') as file:
        mock_deals = json.load(file)

    mock_getThreads.return_value = mock_deals
    dealabs = Dealabs()

    for deal in dealabs.getDeals(False):
        # in Title
        if deal.id == 2496442:
            assert deal.isFree() == True

        # in Price
        elif deal.id == 2496436:
            assert deal.isFree() == True

        # in Groups
        elif deal.id == 2496439:
            assert deal.isFree() == True


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_fields(mock_getThread):
    with open('tests/mocks/deal_fields.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    check_deal_fields(deal)

    assert deal.priceReduction == 170
    assert deal.priceDiscount == 94.44
    assert deal.getHash() == "5200926d6bf01b884013ef7cf4013561"
    assert deal.image is not None
    assert deal.redirectUrl is not None


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_deactivated(mock_getThread):
    with open('tests/mocks/deal_deactivated.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isDeactivated() == True


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_location_true(mock_getThread):
    with open('tests/mocks/deal_location_true.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isLocal() == True


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_location_false(mock_getThread):
    with open('tests/mocks/deal_location_false.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isLocal() == False


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_location_false2(mock_getThread):
    with open('tests/mocks/deal_location_false2.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isLocal() == False


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_error_false(mock_getThread):
    with open('tests/mocks/deal_error_false.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isError() == False


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_error_true_title(mock_getThread):
    with open('tests/mocks/deal_error_true_title.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isError() == True


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_error_true_desc(mock_getThread):
    with open('tests/mocks/deal_error_true_desc.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isError() == True


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_error_true_group(mock_getThread):
    with open('tests/mocks/deal_error_true_group.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isError() == True


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_free_false(mock_getThread):
    with open('tests/mocks/deal_free_false.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isFree() == False


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_free_true_title(mock_getThread):
    with open('tests/mocks/deal_free_true_title.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isFree() == True


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_free_true_price(mock_getThread):
    with open('tests/mocks/deal_free_true_price.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isFree() == True


@patch('dealabs.Dealabs._getThread')
def test_getDeal_check_free_true_group(mock_getThread):
    with open('tests/mocks/deal_free_true_group.json') as file:
        mock_deal = json.load(file)

    mock_getThread.return_value = mock_deal
    dealabs = Dealabs()

    deal = dealabs.getDeal(0)
    assert deal.isFree() == True


dealabs/utils.py
from datetime import datetime
import requests
import os
import json
import re


class Utils:

    # Envoyer le deal par notification
    @staticmethod
    def sendNotification(token, user, deal, open_dealabs=1, priority_only_first=1, free_products_priority=0):
        notif_title = deal.formatTitle()
        urlRedirection = Utils.getFinalUrl(deal.redirectUrl)
        addedModeration = "1" if deal.isAddedInModeration() else "0"
        notif_message = "<b>{}</b>\n\n\n\n{} - <a href='{}'>Voir le site</a> - <a href='{}'>Voir le deal</a>\n{}".format(deal.formatTitle(True), addedModeration, urlRedirection, deal.url, Utils.removeHTML(deal.description))

        url = deal.url if open_dealabs else urlRedirection

        priority = 1

        # Si le deal est gratuit mais qu'on ne veut pas de notifications prioritaires
        if deal.isFree() and free_products_priority == 0:
            priority = 0

        # Si le deal a déjà été envoyé et que le paramètre est activé, on met la priorité à 0
        if deal.id in Utils.getDealSends():
            if priority_only_first:
                priority = 0
        else:
            Utils.addDealSend(deal.id)

        # Post to Pushover
        response = requests.post("https://api.pushover.net/1/messages.json", data={
            "token": token,
            "user": user,
            "title": notif_title,
            "message": notif_message,
            "html": 1,
            "url": url,
            "url_title": "Voir le deal",
            "priority": priority
        }, files={
            "attachment": ("image.jpg", Utils.getImage(deal.image), "image/jpeg")
        }).json()

        if "status" in response and response['status'] == 1:
            Utils.log("[Send] {} [{}]".format(deal.formatTitle(True), str(priority)))
        else:
            Utils.log("[Error] {} [{}]".format(deal.formatTitle(True), str(priority)))

    # Récuperer l'image sous forme de bytes
    @ staticmethod
    def getImage(url):
        headers = {
            'Host': 'static-pepper.dealabs.com',
            'accept': 'image/*',
            'user-agent': 'Dealabs/612302 CFNetwork/1120 Darwin/19.0.0',
            'accept-language': 'fr-fr',
        }

        try:
            response = requests.get(
                url,
                headers=headers,
                timeout=5,
            ).content
        except:
            response = None

        return response

    # Récupérer le répertoire de cache (le créer si il n'existe pas)
    @ staticmethod
    def getCacheFolder():
        if not os.path.exists('.cache'):
            os.makedirs('.cache')
        return '.cache/'

    # Récupérer le fichier blacklist.txt et le retourne sous forme de liste
    @ staticmethod
    def getBlackListFile():
        if os.path.exists('blacklist.txt'):
            with open('blacklist.txt', 'r') as f:
                return f.read().splitlines()
        else:
            return []

    # Récupérer les deals déjà vus
    @ staticmethod
    def getDealsViewed():
        try:
            with open(Utils.getCacheFolder() + 'deals.json', 'r') as f:
                deals = json.load(f)
        except:
            deals = []

        return deals

    # Récupérer les deals déjà notifiés
    @ staticmethod
    def getDealSends():
        try:
            with open(Utils.getCacheFolder() + 'sends.json', 'r') as f:
                deals = json.load(f)
        except:
            deals = []

        return deals

    # Ajouter un deal à la liste des deals vus
    @ staticmethod
    def addDealViewed(hash):
        dealsViewed = Utils.getDealsViewed()
        dealsViewed.append(hash)
        with open(Utils.getCacheFolder() + 'deals.json', 'w') as f:
            json.dump(dealsViewed, f)

    # Ajouter un deal à la liste des deals notifiés
    @ staticmethod
    def addDealSend(id):
        dealsSends = Utils.getDealSends()
        dealsSends.append(id)
        with open(Utils.getCacheFolder() + 'sends.json', 'w') as f:
            json.dump(dealsSends, f)

    # Envoyer une notification pour prévenir que les tokens vont bientôt expirer
    @ staticmethod
    def expireNotification(token, user, refresh_seconds=0):
        try:
            delta_time = int(datetime.now().timestamp() - os.path.getmtime('.env'))
        except FileNotFoundError:
            return

        if delta_time > 29 * 24 * 60 * 60 and delta_time < 29 * 24 * 60 * 60 + refresh_seconds:
            requests.post("https://api.pushover.net/1/messages.json", data={
                "token": token,
                "user": user,
                "title": "🚨 Expiration PushOver",
                "message": "Attention vos tokens expirent demain !",
                "priority": 1
            })

    # Log dans la console
    @ staticmethod
    def log(message):
        print("[{}] {}".format(datetime.now().strftime("%Y-%m-%d %H:%M:%S"), message))

    # Formater le prix
    @ staticmethod
    def formatPrice(price):
        return "{:.2f}€" .format(price)

    # Supprimer les balises html d'une chaine de caractères
    @ staticmethod
    def removeHTML(str):
        return re.sub(re.compile('<.*?>'), '', str).strip()

    # Récupérer l'url finale (après redirections)
    @ staticmethod
    def getFinalUrl(url):
        try:
            response = requests.get(url, headers={
                'user-agent': 'Dealabs/612302 CFNetwork/1120 Darwin/19.0.0'
            }, timeout=5, allow_redirects=True)
            return response.url
        except:
            return url


main.py
import os
import sys
from dealabs import Dealabs, Utils
from dotenv import load_dotenv
load_dotenv()


def help():
    print("""Usage: python3 main.py [refresh_seconds] [minimum_discount] [free_products] [expire_notification] [open_dealabs] [priority_only_first] [free_products_priority]
- refresh_seconds: Nombre de secondes entre chaque rafraichissement
- minimum_discount: Pourcentage minimum de réduction pour être notifié
- free_products: Être notifié des produits gratuits
- expire_notification: Recevoir une notification avant que les tokens pushover ne soient expirés
- open_dealabs: Ouvrir dealabs au lieu du site du deal
- priority_only_first: Notifier uniquement en priorité a l'apparition du deal et pas à chaque mise à jour
- free_products_priority: Envoi également les notifications des produits gratuit en priorité
""")
    sys.exit(0)


def main(args):
    if os.getenv("PUSHOVER_TOKEN") is None:
        print("PUSHOVER_TOKEN not found in .env file")
        exit(1)

    if os.getenv("PUSHOVER_USER") is None:
        print("PUSHOVER_USER not found in .env file")
        exit(1)

    if len(args) > 0 and "h" in args[0]:
        help()

    # Prend les arguments donnés en priorité puis les variables d'environnement, sinon prend les valeurs par défaut
    if len(args) > 0:
        refresh_seconds = int(args[0])
    else:
        refresh_seconds = int(os.getenv("REFRESH_SECONDS")) if os.getenv("REFRESH_SECONDS") else 60

    if len(args) > 1:
        minimum_discount = int(args[1])
    else:
        minimum_discount = int(os.getenv("MIN_DISCOUNT")) if os.getenv("MIN_DISCOUNT") else 90

    if len(args) > 2:
        free_products = int(args[2])
    else:
        free_products = int(os.getenv("FREE_PRODUCTS")) if os.getenv("FREE_PRODUCTS") else 0

    if len(args) > 3:
        expire_notification = int(args[3])
    else:
        expire_notification = int(os.getenv("EXPIRE_NOTIFICATION")) if os.getenv("EXPIRE_NOTIFICATION") else 0

    if len(args) > 4:
        open_dealabs = int(args[4])
    else:
        open_dealabs = int(os.getenv("OPEN_DEALABS")) if os.getenv("OPEN_DEALABS") else 1

    if len(args) > 5:
        priority_only_first = int(args[5])
    else:
        priority_only_first = int(os.getenv("PRIORITY_ONLY_FIRST")) if os.getenv("PRIORITY_ONLY_FIRST") else 1

    if len(args) > 6:
        free_products_priority = int(args[6])
    else:
        free_products_priority = int(os.getenv("FREE_PRODUCTS_PRIORITY")) if os.getenv("FREE_PRODUCTS_PRIORITY") else 0

    dealabs = Dealabs(minimum_discount, free_products, expire_notification, open_dealabs, priority_only_first, free_products_priority)
    dealabs.setPushOver(os.getenv("PUSHOVER_TOKEN"), os.getenv("PUSHOVER_USER"))
    dealabs.watch(refresh_seconds)


if __name__ == "__main__":
    try:
        main(sys.argv[1:])
    except KeyboardInterrupt:
        pass
